MGMT-22553: Partial cluster updates fail with false validation errors on dual-stack clusters

LiorSoffer opened the pull request:
https://issues.redhat.com/browse/MGMT-22553

**Root Cause**

The ValidateClusterUpdateVIPAddresses function assigned network and VIP parameters directly from the update request without falling back to the cluster's existing values when parameters were nil.
VIPs were initialized from `params.APIVips` and `params.IngressVips` directly, so if only one was provided, the other was treated as empty (length 0) instead of using the cluster's existing value

Network configurations (ClusterNetworks, ServiceNetworks, MachineNetworks) were assigned from params without checking for nil, so unspecified networks became empty during validation instead of preserving cluster's existing values.

**The Fix**
1. Use cluster's existing APIVips/IngressVips as defaults, override only when user provides new values in the update request
2. Networks - Fallback to cluster values when params are nil:
If `params.ClusterNetworks` is nil, use `cluster.ClusterNetworks`
If `params.ServiceNetworks` is nil, use `cluster.ServiceNetworks`
If `params.MachineNetworks` is nil, use `cluster.MachineNetworks`
3. DHCP mode handling - Refactored for clarity:
Only clear VIPs when explicitly switching to DHCP mode AND user didn't provide VIPs
If user sends VIPs + DHCP=true, let validation fail with appropriate error message

**Changes**

internal/cluster/validations/validations.go - Fixed fallback logic for VIPs and network configurations
internal/cluster/validations/validation_test.go - Added test cases for partial network updates on dual-stack clusters
internal/bminventory/inventory_test.go - Fixed test data consistency and added VIPs to dual-stack test params

omer-vishlitzky commented:
/test

omer-vishlitzky commented:
/test e2e-agent-ha-dualstack

omer-vishlitzky commented:
/test edge-e2e-vsphere-assisted-umn-4-20

omer-vishlitzky commented:
/test edge-e2e-metal-assisted-static-ip-suite-4-20

omer-vishlitzky commented:
/test edge-e2e-metal-assisted-ipv4v6-4-20

omer-vishlitzky commented:
/retest

omer-vishlitzky commented:
/test e2e-agent-ha-dualstack edge-e2e-vsphere-assisted-umn-4-20 edge-e2e-metal-assisted-static-ip-suite-4-20 edge-e2e-metal-assisted-ipv4v6-4-20 edge-e2e-metal-assisted-dual-stack-primary-ipv6-4-20 edge-e2e-metal-assisted-dual-primary-v6-compact-4-20

LiorSoffer commented:
@linoyaslan @carbonin   sorry for the lack of detail , I've updated both the JIRA ticket and PR description with a clearer explanation of the bug and fix :)

linoyaslan commented on internal/cluster/validations/validations.go:
```diff
@@ -264,8 +264,23 @@ func ValidateClusterUpdateVIPAddresses(ipV6Supported bool, cluster *common.Clust
+	// Determine if we're switching to a mode where VIPs are not allowed
+	switchingToUMN := swag.BoolValue(params.UserManagedNetworking)
+	switchingToDHCP := swag.BoolValue(params.VipDhcpAllocation)
+	preserveVIPs := !switchingToUMN && !switchingToDHCP
```
what if a user initially provided VIPs but later wants to explicitly remove them? How is that supported with this approach? It seems like this would always end up with an error, unless I’m missing something here?

LiorSoffer replied:
if I understand correctly, in such cases the user must expicitly remove them before switiching to UMN

linoyaslan replied:
Never mind @omer-vishlitzky  and I went through the code and found that the VIPs are reset when switching to UMN, so my comment isn't relevant.

LiorSoffer committed:
MGMT-22553: Fix VIP validation for partial cluster updates

- Validation function uses cluster's current values as defaults when
  update params don't explicitly provide them (needed for partial updates)
- Clear VIPs when user toggles VipDhcpAllocation mode to prevent:
  1. Non-DHCP mode using VIPs previously obtained via DHCP
  2. DHCP mode being fed with manually provided VIPs
  (If user explicitly provides VIPs while changing DHCP mode,
  validation will fail appropriately)

omer-vishlitzky commented on internal/cluster/validations/validations.go:
```diff
@@ -258,14 +258,17 @@ func validateVIPsWithUMA(cluster *common.Cluster, params *models.V2ClusterUpdate
+	}
+	if params.IngressVips != nil {
+		ingressVips = params.IngressVips
+	}
```
intentional - we only clear the vips if we set them up in the first place (dhcp mode). in all other cases the user is required to explicitly clear them.

CrystalChun commented:
/cherry-pick release-ocm-2.15
